2.train-VAE/betatcvae.py

```diff
@@ -0,0 +1,300 @@
+        return recon_x, mu, logvar, z
+
+    def log_density_gaussian(self, x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor):
+        normalization = -0.5 * (math.log(2 * math.pi) + logvar) 
+        inv_var = torch.exp(-logvar)  
+        log_density = normalization - 0.5 * ((x - mu)**2 * inv_var)  
+        return log_density
+
+    def loss_function(self,
+                      *args,
+                      **kwargs) -> dict:
+        recons = args[0]  
+        input = args[1]  
+        mu = args[2]  
+        log_var = args[3]  
+        z = args[4] 
+        M_N = kwargs.get('M_N', 1.0)  # Minibatch size
+
+        weight = 1  # kwargs['M_N']  # Account for the minibatch samples from the dataset
+
+        recons_loss = F.mse_loss(recons, input, reduction='sum')  
+
+        log_q_zx = self.log_density_gaussian(z, mu, log_var).sum(dim=1)  # Log q(z|x)
+
+        zeros = torch.zeros_like(z)
+        log_p_z = self.log_density_gaussian(z, zeros, zeros).sum(dim=1)  # Log p(z)
+
+        batch_size, latent_dim = z.shape
+        mat_log_q_z = self.log_density_gaussian(z.view(batch_size, 1, latent_dim),
+                                                mu.view(1, batch_size, latent_dim),
+                                                log_var.view(1, batch_size, latent_dim))  # Log q(z)
+
+        dataset_size = (1 / M_N) * batch_size  # Dataset size
+        strat_weight = (dataset_size - batch_size + 1) / (dataset_size * (batch_size - 1))  # Stratified weight
+        importance_weights = torch.Tensor(batch_size, batch_size).fill_(1 / (batch_size - 1)).to(input.device)
+        importance_weights.view(-1)[::batch_size] = 1 / dataset_size  # Importance weights
+        importance_weights.view(-1)[1::batch_size] = strat_weight
+        importance_weights[batch_size - 2, 0] = strat_weight
+        log_importance_weights = importance_weights.log()  # Log importance weights
+
+        mat_log_q_z += log_importance_weights.view(batch_size, batch_size, 1)  # Adjusted log q(z)
+
+        log_q_z = torch.logsumexp(mat_log_q_z.sum(2), dim=1, keepdim=False)  # Log q(z)
+        log_prod_q_z = torch.logsumexp(mat_log_q_z, dim=1, keepdim=False).sum(1)  # Log product of q(z)
+
+        mi_loss = (log_q_zx - log_q_z).mean()  # Mutual information loss
+        tc_loss = (log_q_z - log_prod_q_z).mean()  # Total correlation loss
+        kld_loss = (log_prod_q_z - log_p_z).mean()  # KL divergence loss
+
+        if self.training:
+            self.num_iter += 1  # Increment iteration count if training
+            anneal_rate = min(0 + 1 * self.num_iter / self.anneal_steps, 1)  # Anneal rate
+        else:
+            anneal_rate = 1.  # No annealing if not training
+
+        # Total loss with weighted components
+        loss = recons_loss + (self.alpha * mi_loss +  
+            self.beta * tc_loss +
+            anneal_rate * self.gamma * kld_loss)
+        
+        return loss
+
+
+
+
+    
+def train_model(model, train_loader, optimizer):
```

Comment:
Also consider splitting these up into different `.py` files. I could see a `tcvae_model.py` file for the class and a `tcvae_train_utils.py` file for the training functions.